search-index/text2embeddings.py

```diff
@@ -18,10 +19,108 @@
 logger = get_logger(__name__)
 
 
+def get_text_from_list(text_block: dict, prev_processed_text_block: dict) -> str:
+    """
+    Format a list text block and prepend it to the previous text block, assuming it is the context of the list.
+    This is a fairly strong assumption, but works most of the time. Code to handle more complex cases has been started.
+
+    Args:
+        text_block (dict): text block dict.
+        prev_processed_text_block (dict): previous text block dict.
+
+    Returns:
+        str: Formatted list text block that is useful for indexing semantics.
+    """
+    text = text_block["custom_attributes"]["pretty_list_string"]
+    text = text.strip()
+    text = text.replace("*", "")  # Remove bullet points.
+    # Always append previous text block as context. This is a fairly strong assumption, and the code to make this
+    # better has been started, in postprocessor.py, but is not yet completely robust, so we are not using it for now.
+    # TODO: Note, we can potentially do this in a more robust way by using the text block custom attributes created
+    #  in postprocessor.py, which I've updated. This will allow us to deal with edge cases, for example where the
+    #  previous block is not either context or part of the same list from the previous page. This would be the case
+    #  when, for example, there are random elements between pages that the postprocessors haven't been able to catch.
+    #  But I've ignored this for now because it seems that the postprocessing in postprocessor.py has dealt with the
+    #  vast majority of edge cases (as indicated by the rarity of certain debugging metadata).
+    text = prev_processed_text_block + "\n" + text
+    return text
+
+
+def delete_string_indices(data: str, indices: List[int]) -> str:
+    """Delete a list of indices from a string.
+    Args:
+        data (str): string to delete indices from.
+        indices (List[int]): list of indices to delete.
+    Returns:
+        str: string with indices deleted.
+    """
+    return "".join([char for idx, char in enumerate(data) if idx not in indices])
+
+
+def get_text_from_merged_block(text_block: dict) -> str:
+    """
+    Remove unnecessary styling from merged text blocks (superscripts).
+
+    Args:
+        text_block: A block of text that has been merged.
+
+    Returns:
+        str: The text block with styling that is not related to semantics removed.
+
+    """
+    # Text output with no processing.
+    text_output = (
+        "\n".join(text_block["text"]).strip(),
```

Comment:
I think we want to actually `" ".join()` strings here so that displaying them in the UI doesn't enforce arbitrary line breaks. We should still newline join where newlines are part of the semantics, e.g. for lists or paragraph breaks. This is my mistake from the original method that I've spotted since. What do you think?